lib/src/kbasesearchengine/common/GUID.java

```diff
@@ -128,21 +152,54 @@ public GUID(String textGUID) {
             this.subObjectId = subObjectPart.substring(slashPos + 1);
         }
     }
-    
-    //TODO NOW if the version or accessGroupID is missing, how do you know which is which? Need to know storage code -> has access group mapping or have a placeholder (-1) for missing access group
+
+    /**
+     * Returns the URL-encoded version of this GUID.
+     *
+     * For the sake of readability the URL-encoded format of the GUID is,
+     *
+     *   {storageCode}___[{AccessGroupID}_]{accessGroupObjectId}[_{version}][___{subObjectType}_{subObjectId}],
+     *
+     *  where special characters in the individual tokens will get URL-encoded.
+     *
+     * @return URL-encoded version of this GUID
+     */
+    public String getURLEncoded() throws IOException {
+
+        String encodedGUID = toString("___", "_");
```

Comment:
What's the reasoning for swapping ":" for "___"?

Reply:
":" encodes to "%3A" and makes the guid look ugly.

Reply:
But why do you need to encode it at all? It should work fine without encoding even with curl:
```
crusherofheads@icrushdeheads:~$ curl "http://homology.kbase.us/namespace/KBase:Refseq" | python -m json.tool
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   248  100   248    0     0   1578      0 --:--:-- --:--:-- --:--:--  1579
{
    "error": {
        "appcode": 30001,
        "apperror": "Illegal input parameter",
        "callid": "4105873853039241",
        "httpcode": 400,
        "httpstatus": "Bad Request",
        "message": "30001 Illegal input parameter: Illegal character in namespace id KBase:Refseq: :",
        "time": 1527730889932
    }
}
```

Reply:
In your example the output states that the request was bad.

Reply:
But that's the business layer talking, not the http layer. The business layer got a colon, not the encoding for a colon.

Reply:
This is what I think is happening. ":" is a valid but reserved char in http. Since it is valid, http will not complain. But once this info is passed on to the receiving application, the info make no sense to the application because "KBase:Refseq" is probably being broken up into host segment and port by the http layer and then being supplied to the app layer. So the app layer complains as it doesn't understand the request.

Reply:
No, it's received as a single string.

Reply:
Then I guess it's because the app doesn't expect a ":" in a path segment. Not part of its API spec.

Reply:
It's because a colon is an illegal character in a namespace ID.

Reply:
Makes sense. I was wrong, the port ":" comes much before the path segments.

Reply:
Since ES API doesn't have a problem with ":" in path segment I have switched "___" delim with ":". 
Here's what the _id looks like now - 
WS:1_1_1:contig_NZ_MCBT01000008